flashlight/fl/tensor/backend/jit/opt/passes/ScalarFolding.cpp

```diff
@@ -0,0 +1,115 @@
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "flashlight/fl/tensor/backend/jit/opt/passes/ScalarFolding.h"
+
+#include "flashlight/fl/tensor/backend/jit/ir/BinaryNode.h"
+#include "flashlight/fl/tensor/backend/jit/ir/ScalarNode.h"
+
+namespace fl {
+
+namespace {
+
+template <typename T>
+T foldScalars(const T lhs, const T rhs, const BinaryOp op) {
+  switch (op) {
+    case BinaryOp::Add:
+      return lhs + rhs;
+    case BinaryOp::Sub:
+      return lhs - rhs;
+    case BinaryOp::Mul:
+      return lhs * rhs;
+    case BinaryOp::Div:
+      return lhs / rhs;
+  }
+  throw std::runtime_error("[foldScalars] Unknown binary operation type");
+}
+
+template <typename T>
+ScalarNode* foldScalarNodes(
+    const ScalarNode& lhs,
+    const ScalarNode& rhs,
+    const BinaryOp op,
+    const dtype type) {
+  T lhsVal = lhs.scalar<T>();
+  T rhsVal = rhs.scalar<T>();
+  T resVal = foldScalars(lhsVal, rhsVal, op);
+  return ScalarNode::create(Shape(lhs.shape()), type, resVal);
+}
+
+ScalarNode* foldScalarNodes(
+    const ScalarNode& lhs,
+    const ScalarNode& rhs,
+    const BinaryOp op,
+    const dtype type) {
+  switch (type) {
+    case dtype::f16:
+      throw std::runtime_error("[foldScalarNodes] f16 is unsupported");
+    case dtype::f32:
+      return foldScalarNodes<float>(lhs, rhs, op, type);
+    case dtype::f64:
+      return foldScalarNodes<double>(lhs, rhs, op, type);
+    case dtype::b8:
+      return foldScalarNodes<char>(lhs, rhs, op, type);
+    case dtype::s16:
+      return foldScalarNodes<short>(lhs, rhs, op, type);
+    case dtype::s32:
+      return foldScalarNodes<int>(lhs, rhs, op, type);
+    case dtype::s64:
+      return foldScalarNodes<long long>(lhs, rhs, op, type);
+    case dtype::u8:
+      return foldScalarNodes<unsigned char>(lhs, rhs, op, type);
+    case dtype::u16:
+      return foldScalarNodes<unsigned short>(lhs, rhs, op, type);
+    case dtype::u32:
+      return foldScalarNodes<unsigned int>(lhs, rhs, op, type);
+    case dtype::u64:
+      return foldScalarNodes<unsigned long long>(lhs, rhs, op, type);
+  }
+  throw std::runtime_error("[foldScalarNodes] Unknown data type");
+}
+
+Node* foldScalarsInBinaryNode(BinaryNode* node) {
+  const auto binop = node->op();
+  const auto lhs = node->lhs();
+  const auto rhs = node->rhs();
+  if (lhs->isScalar() && rhs->isScalar()) {
+    const auto& lhsScalar = lhs->impl<ScalarNode>();
+    const auto& rhsScalar = rhs->impl<ScalarNode>();
+    const auto& shape = lhsScalar.shape();
+    const auto dtype = lhsScalar.dataType();
+    if (shape == rhsScalar.shape() && dtype == rhsScalar.dataType()) {
+      auto foldedScalar = foldScalarNodes(lhsScalar, rhsScalar, binop, dtype);
+      node->replaceAllUsesWith(foldedScalar);
+      return foldedScalar;
+    }
+  }
+  return node;
+}
+
+Node* foldScalars(Node* node) {
+  for (const auto& input : node->inputs()) {
+    foldScalars(input);
```

Comment:
Just curious -- why apply this optimization in a prefix way? Makes sense that you're folding all of the inputs, but what if you wanted to apply other optimizations to the inputs?

Reply:
> what if you wanted to apply other optimizations to the inputs?

Wouldn't that be another pass, since this is just a scalar-folding pass?

So it's a prefix-traversal because we want to reduce nested scalar-binops in 1 single pass, e.g.,

```
2   3
 \  /
  *   4
   \ /
    +
```

Reply:
👍